### PR TITLE
SDK progress

### PR DESCRIPTION
**Pull request by pharmapsychotic**
* cv2 -> PIL
* Auto routing to transform-server-v1-cpu engine
* Unit tests for transforms and chains
* Upscale support for animations
* Clean ups and merge preps

### CODE REVIEW

**Review by aletts — APPROVED** [review]

File: src/stability_sdk/animation.py [hunk 1]
```diff
@@ -270,25 +271,26 @@ def interpolate_frames(
...
 
-def mask_erode_blur(mask: np.ndarray, mask_erode: int, mask_blur: int) -> np.ndarray:
+def mask_erode_blur(mask: Image.Image, mask_erode: int, mask_blur: int) -> Image.Image:
+    mask = np.array(mask)
     if mask_erode > 0:
         ks = mask_erode*2 + 1
         mask = cv2.erode(mask, cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (ks, ks)), iterations=1)
     if mask_blur > 0:
         ks = mask_blur*2 + 1
         mask = cv2.GaussianBlur(mask, (ks, ks), 0)
```
Comment (kostarion) [inline src/stability_sdk/animation.py:292]: It's to be replaced with PIL.ImageFilter.GaussianBlur() and PIL.ImageFilter.MinFilter(), right?

File: src/stability_sdk/animation_ui.py [hunk 2]
```diff
@@ -271,43 +281,82 @@ def postprocess_video(fps: int, reverse: bool, interp_mode: str, interp_factor:
...
+                            process_button: gr.update(visible=False),
+                            stop_button: gr.update(visible=True),
+                            error_log: gr.update(visible=False),
+                        }
+                        if interrupt:
+                            break
+                    last_interp_mode, last_interp_factor = interp_mode, interp_factor
+                outdir = interp_dir
+
+            output_video = last_project_settings_path.replace(".json", f"{suffix}.mp4")
```
Comment (kostarion) [inline src/stability_sdk/animation_ui.py:341]: cool! yesterday i was struggling to find the postprocessed video in the project folder so I had to click through all mp4. glad it's easier now!

File: src/stability_sdk/api.py [hunk 3]
```diff
@@ -7,15 +8,6 @@
 from PIL import Image
 from typing import Any, Dict, Iterable, List, Optional, Sequence, Tuple, Union
```
Comment (kostarion) [inline src/stability_sdk/api.py:9]:
```suggestion
from typing import Any, Dict, List, Optional, Sequence, Tuple, Union
```

File: src/stability_sdk/utils.py [hunk 4]
```diff
@@ -57,84 +27,169 @@
...
+    "k_dpmpp_2m": generation.SAMPLER_K_DPMPP_2M,
+    "k_dpmpp_2s_ancestral": generation.SAMPLER_K_DPMPP_2S_ANCESTRAL
+}
 
-def cv2_to_pil(cv2_img: np.ndarray) -> Image.Image:
-    """Convert a cv2 BGR ndarray to a PIL Image"""
-    return Image.fromarray(cv2_img[:, :, ::-1])
+T = TypeVar('T')
 
+def _from_string(s: str, mapping: Dict[str, T], name: str, enum_cls: Type[T]) -> T:
```
Comment (kostarion) [inline src/stability_sdk/utils.py:76]: 🎉 thank you! it haunted me for months :D

**Review by kostarion — APPROVED** [review]
awesome pr! thank you
i think it should be merged with my recent changes, replace last cv2 bits, but overall good to go

File: src/stability_sdk/animation.py [hunk 5]
```diff
@@ -270,25 +271,26 @@ def interpolate_frames(
...
-            cv2.imwrite(os.path.join(out_path, f"frame_{i * interp_factor + ti + 1:05d}.png"), tween)
-            yield cv2_to_pil(tween)
+            tween.save(os.path.join(out_path, f"frame_{i * interp_factor + ti + 1:05d}.png"))
+            yield tween
 
     # copy final frame
     shutil.copy(frame_files[-1], os.path.join(out_path, f"frame_{(len(frame_files)-1) * interp_factor:05d}.png"))        
 
-def mask_erode_blur(mask: np.ndarray, mask_erode: int, mask_blur: int) -> np.ndarray:
+def mask_erode_blur(mask: Image.Image, mask_erode: int, mask_blur: int) -> Image.Image:
```
[the remark this answers is not in the document]
Comment (pharmapsychotic) [reply src/stability_sdk/animation.py:285]: yeah good call, I'll merge up then eliminate it!

File: src/stability_sdk/animation.py [hunk 6]
```diff
@@ -93,7 +92,7 @@ class AnimationSettings(param.Parameterized):
     noise_scale_curve = param.String(default="0:(0.99)")
     strength_curve = param.String(default="0:(0.65)", doc="Image Strength (of init image relative to the prompt). 0 for ignore init image and attend only to prompt, 1 would return the init image unmodified")
     steps_curve = param.String(default="0:(30)", doc="Diffusion steps")
-    steps_strength_adj = param.Boolean(default=True, doc="Adjusts number of diffusion steps based on current previous frame strength value.")    
+    steps_strength_adj = param.Boolean(default=False, doc="Adjusts number of diffusion steps based on current previous frame strength value.")
```
[the remark this answers is not in the document]
Comment (pharmapsychotic) [reply src/stability_sdk/animation.py:95]: This hits pricing target and with default steps 30 I had found poor results with `steps_strength_adj` on with SDXL. Added to follow up list of checklist to evaluate and consider further.

File: src/stability_sdk/animation.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (pharmapsychotic) [reply src/stability_sdk/animation.py:286]: Will hold for later change, they aren't quite compatible yet and should update inpaint test notebook

**Pull request merged**